mesh/nurbs.hpp

```diff
@@ -30,60 +30,118 @@ namespace mfem
 class GridFunction;
 
 
+/** @brief A vector of knots in one dimension, with B-spline basis functions of
+    a prescribed order.
+
+    @note Order is defined in the sense of "The NURBS book" - 2nd ed - Piegl and
+    Tiller, cf. section 2.2.
+*/
 class KnotVector
 {
 protected:
    static const int MaxOrder;
 
-   Vector knot; // Values of knots
-   int Order, NumOfControlPoints, NumOfElements;
+   /// Stores the values of all knots.
+   Vector knot;
+
+   /// Order of the B-spline basis functions.
+   int Order;
+
+   /// Number of control points.
+   int NumOfControlPoints;
+
+   /// Number of elements, defined by distinct knots.
+   int NumOfElements;
 
 public:
-   /// Create KnotVector
+   /// Create an empty KnotVector.
    KnotVector() { }
+
+   /// Create a KnotVector by reading data from stream @a input.
    KnotVector(std::istream &input);
-   KnotVector(int Order_, int NCP);
+
+   /** @brief Create a KnotVector with undefined knots (initialized to -1) of
+       order @a order and number of control points @a NCP. */
+   KnotVector(int order, int NCP);
+
+   /// Copy constructor.
    KnotVector(const KnotVector &kv) { (*this) = kv; }
 
    KnotVector &operator=(const KnotVector &kv);
 
+   /// Return the number of elements, defined by distinct knots.
    int GetNE()    const { return NumOfElements; }
-   int GetNKS()   const { return NumOfControlPoints - Order; }
+
+   /// Return the number of control points.
    int GetNCP()   const { return NumOfControlPoints; }
+
+   /// Return the order.
    int GetOrder() const { return Order; }
+
+   /// Return the number of knots, including multiplicities.
    int Size()     const { return knot.Size(); }
 
-   /// Count the number of elements
+   /// Count the number of elements.
    void GetElements();
 
+   /** @brief Return whether the knot index Order plus @a i is the beginning of
+       an element. */
    bool isElement(int i) const { return (knot(Order+i) != knot(Order+i+1)); }
 
+   /// Return the number of control points minus the order.
```

Comment:
Is this the number of knots?

Reply:
The short answer is no, but let me try to explain based on "The NURBS Book". This could be a helpful piece of documentation, but I'm not sure where in the code it should go. Maybe at the top of nurbs.hpp? Or maybe we should have a page on the MFEM website giving an overview of NURBS in MFEM.

A p-th degree (order p+1, cf. section 2.2) B-spline curve has

n + 1 control points,
m + 1 knots

with the first p+1 knots being 0 and the last p+1 knots being 1. Interior knots may be repeated (multiplicity greater than 1). 

The piecewise polynomial degree is p. In the code, `Order`, returned by `GetOrder`, is this degree p (cf. `KnotVector::CalcShape`). Thus we use degree and order interchangeably and equally, although the book has order = p + 1. Maybe we should change this in MFEM 5.0, by having a breaking change to use "order" and "degree" correctly.

P3.2 of the book gives the relation

m = n + p + 1

A knot span is considered to be non-empty, meaning it must start at the last repetition of a knot. With this understanding of knot span, the number of knot spans is not given by a simple formula, since knot multiplicities must be checked. That is why multiple functions in nurbs.cpp loop over `GetNKS()` indices and call `isElement` to check for a non-empty knot span. Therefore, `GetNKS()` is not actually the number of knot spans. Instead, it simply computes
`NumOfControlPoints - Order`
or
n + 1 - p = m - 2p
in the notation of the NURBS book.
The number of knots is m + 1. When looping over knot spans, we skip the first p knots and last p+1 knots, meaning we only need to check
m + 1 - p - p - 1 = m - 2p
knots, which is `GetNKS()`.

The current documentation of `GetNKS` is not very helpful: `number of control points minus the order.` However, I'm not sure how to give a more insightful description of this function without writing paragraphs.

Reply:
I think "number of knots that could possibly be the beginning of a knot span" is probably better than what is there.  Maybe that and a sentence about the most likely use case?

Reply:
I expanded the comment to explain how it is used.

Reply:
Thanks, that will probably avoid some head scratching.